src/next_array.rs

```diff
@@ -0,0 +1,64 @@
+use core::mem::MaybeUninit;
+use core::ptr;
+
+/// Helper struct to build up an array element by element.
+struct ArrayBuilder<T, const N: usize> {
+    arr: [MaybeUninit<T>; N], // Invariant: arr[..len] is initialized.
+    len: usize,               // Invariant: len <= N.
+}
+
+impl<T, const N: usize> ArrayBuilder<T, N> {
+    pub fn new() -> Self {
+        Self {
+            arr: [(); N].map(|_| MaybeUninit::uninit()),
+            len: 0,
+        }
+    }
+
+    pub fn push(&mut self, value: T) {
+        // We maintain the invariant here that arr[..len] is initialized.
+        // Indexing with self.len also ensures self.len < N, and thus <= N after
+        // the increment.
+        self.arr[self.len] = MaybeUninit::new(value);
+        self.len += 1;
+    }
+
+    pub fn take(&mut self) -> Option<[T; N]> {
+        if self.len == N {
+            // Take the array, resetting the length back to zero.
+            let arr = core::mem::replace(&mut self.arr, [(); N].map(|_| MaybeUninit::uninit()));
+            self.len = 0;
+
+            // SAFETY: we had len == N, so all elements in arr are initialized.
+            Some(unsafe { arr.map(|v| v.assume_init()) })
+        } else {
+            None
+        }
+    }
+}
+
+impl<T, const N: usize> Drop for ArrayBuilder<T, N> {
+    fn drop(&mut self) {
+        unsafe {
+            // SAFETY: arr[..len] is initialized, so must be dropped.
+            // First we create a pointer to this initialized slice, then drop
+            // that slice in-place. The cast from *mut MaybeUninit<T> to *mut T
+            // is always sound by the layout guarantees of MaybeUninit.
+            let ptr_to_first: *mut MaybeUninit<T> = self.arr.as_mut_ptr();
+            let ptr_to_slice = ptr::slice_from_raw_parts_mut(ptr_to_first.cast::<T>(), self.len);
+            ptr::drop_in_place(ptr_to_slice);
+        }
```

Comment:
Since our MSRV is now >= 1.60 we can leverage `MaybeUninit::assume_init_drop`. I know this is contrary to @scottmcm's suggestion [here](https://github.com/rust-itertools/itertools/pull/560#discussion_r1543287242), but it entirely avoids working with raw pointers.

```suggestion
        // LEMMA: The elements of `init` reference the valid elements of
        // `self.arr`.
        //
        // PROOF: `slice::split_at_mut(mid)` produces a pair of slices, the
        // first of which contains the elements at the indices `0..mid`. By
        // invariant on `self.arr`, the elements of `self.arr` at indicies
        // `0..self.len` are valid. Assuming that `slice::split_at_mut` is
        // correctly implemented, the slice `init` will only reference the
        // valid elements of `self.arr`.
        let (init, _) = self.arr.split_at_mut(self.len);
        // AXIOM: We assume that `slice::into_iter` and the subsequent
        // `for_each` are implemented correctly, and will yield each and every
        // element of `init`.
        init.into_iter().for_each(|elt| {
            // SAFETY: `elt` references a valid `T`, because by the above AXIOM,
            // `elt` is an element of `init`, and by the above LEMMA, the
            // elements of `init` reference the valid elements of `self.arr`.
            unsafe { elt.assume_init_drop() }
        });
```

Reply:
This actually has the wrong semantics compared to e.g. `Vec<T>`, as it will leak values if a `Drop` impl panics. `ptr::drop_in_place` will forcibly keep dropping elements in that scenario, causing an abort if another `Drop` panics.

Reply:
> Also, again I find this overly verbose, ultimately reducing readability of the code

I think it is likely that we have different preferences for how unsafe code is written and documented.

Abstractions like the one we're adding here are a notorious source of soundness issues in the Rust ecosystem (see the RUSTSEC advisories for [`smallvec`](https://rustsec.org/packages/smallvec.html), [`stackvector`](https://rustsec.org/packages/stackvector.html), and [`stack`](https://rustsec.org/packages/stack.html), for a sampler). Itertools has not yet had any soundness issues, and I'd like to keep it that way.

I am deeply reluctant to add *any* unsafe code to `itertools`. If we are to do so, I will require that it meets the standards I set for the other unsafe crates I maintain. The safety comments I'm proposing are not just suggestions; they're prerequisites of merging this PR. 

The argument about a semantic difference is interesting. Leaking is not a soundness issue, and we don't make any promises to match the drop behavior of `Vec` under such conditions. Nonetheless, I'll grant that minimizing leaking is preferable. I'll have to give this one more thought.